File: unique-binary-search-tree-2/solution-2.py
from typing import List
from typing import Optional
import copy

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def generate(self, start: int, end:int, records) -> List[Optional[TreeNode]]:
        if start > end:
            return [None]

        if (start, end) in records:
            return records[(start, end)]

        res = []
        for root in range(start, end+1):
            left_trees = self.generate(start, root-1, records)
            right_trees = self.generate(root+1, end, records)

            for left in left_trees:
                for right in right_trees:
                    tree = TreeNode(root)
                    tree.left = left
                    tree.right = right
                    res.append(tree)

        records[(start, end)] = res
        return res

    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        return self.generate(1, n, {})

# help functions
def printTree(tree, outList):
    if tree == None:
        outList.append('null')
        return

    outList.append(tree.val)
    if tree.left != None or tree.right!= None:
        printTree(tree.left, outList)
        printTree(tree.right, outList)

def printTreeList(treeList):
    for tree in treeList:
        outList = []
        printTree(tree, outList)
        print(outList)

if __name__ == '__main__':
    sol = Solution()
    ans = sol.generateTrees(4)
    printTreeList(ans)

File: decode-ways/solution-2.cpp
#include <string>
#include <unordered_set>
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
  int numDecodings(string s) {
    int n = s.size();
    if (n == 0)
      return 0;

    if (s[0] == '0')
      return 0;

    if (n == 1)
      return 1;

    vector<int> dp(n, 0);
    dp[0] = 1;

    if ('1' <= s[1] && s[1] <= '9') {
      dp[1] += dp[0];
    }
    if (s[0] == '1' || (s[0] == '2' && s[1] <= '6')) {
      dp[1] += 1;
    }
    for (int start = 2; start < n; ++start) {
      if ('1' <= s[start] && s[start] <= '9') {
	dp[start] += dp[start - 1];
      }
      if (s[start - 1] == '1' || (s[start - 1] == '2' && s[start] <= '6')) {
	dp[start] += dp[start - 2];
      }
    }

    return dp[n - 1];
  }
};

int main() {
  Solution sol;
  int ret;
  ret = sol.numDecodings("1");     // 1
  cout << ret << endl;
  ret = sol.numDecodings("12");    // 2
  cout << ret << endl;
  ret = sol.numDecodings("226");   // 3
  cout << ret << endl;
  ret = sol.numDecodings("1123");  // 5
  cout << ret << endl;
  ret = sol.numDecodings("06");    // 0
  cout << ret << endl;
  return 0;
}
